### ISSUE TITLE
Multitple 'AssertionFailed' events inserted in inheritance

### ISSUE THREAD

**Issue by cd1m0** [ConsenSys/scribble#132]
Given the following samples:

base.sol:
```
pragma solidity 0.8.7;

/// #invariant true;
contract Base {
    function bar() public {}
}
```

child.sol:
```
pragma solidity 0.8.7;

import "./base.sol";

/// #invariant true;
contract Child is Base {
    function foo() public {}
}
```

scribble produces code that doesn't compile:

```
$ scribble child.sol  base.sol
/// #invariant true;
contract Base is __scribble_ReentrancyUtils {
    event AssertionFailed(string message);
...
}

/// #invariant true;
contract Child is __scribble_ReentrancyUtils, Base {
    event AssertionFailed(string message);
...
}
```

The issue is that both the parent class (Base) and subclass (Child) define the same event `AssertionFailed`.
This error doesn't occur if we switch the order of the arguments, or if we put them in the two contracts in the same file.

The problem stems from the fact that the logic that determines where `AssertionFailed` are instered depends on the order of visiting contracts:

```jsx
export function insertAnnotations(annotations: PropertyMetaData[], ctx: TranspilingContext): void {
...
        const event = getAssertionFailedEvent(factory, contract);
        return [emitAssert(ctx, predicate, annotation, event, emitStmt), false];
...
}

```jsx
export function getAssertionFailedEvent(
    factory: ASTNodeFactory,
    contract: ContractDefinition
): EventDefinition {
    const events = resolveByName(contract, EventDefinition, "AssertionFailed");

    if (events.length > 0) {
        return events[0];
    }
    ...
```

Basically in which-ever contract we first add annotations, we will also emit the event. This is a problem when there are multiple contracts in a single inheritance chain that have the require an event, and when we first happen to visit a child contract.

A cleaner solution would be to always insert `AssertionFailed` and `AssertionFailedData` events in the `__scribble_ReentrancyUtils` contract and make all contracts inherit from it.

**Comment by cd1m0** [comment]
To add more detail: An alternative fix would have been to add a new library with the event and call the event from the library:

```jsx
library ScribbleUtils {
    event AssertionFailed(string message);
}

contract Foo {
    function foo() public {
        emit ScribbleUtils.AssertionFailed("bar");
    }
}
```

Unfortunately this is not supported in solidity < 0.6.2.

If we move `AssertionFailed` to `__scribble_ReentrancyUtils` then we will have to refactor some of the instrumentation logic to
add `__scribble_ReentrancyUtils` as a base for all contracts with annotations. It is currently only added for contracts with `invariant` annotations, and other contracts in their inheritance tree. So now the logic needs to support 2 sets of contracts: the set of contracts that need just the event definition, and the set of contracts that need the reentrancy checks and invariant checks as well.
